diagnostics/diagnostics_manager.rs

```diff
@@ -0,0 +1,155 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{collections::HashSet, future::Future, hash::Hash, sync::Arc};
+
+use error::TypeDBError;
+use resource::constants::{database::INTERNAL_DATABASE_PREFIX, diagnostics::REPORTING_URI};
+use tonic::Status;
+use tonic_types::StatusExt;
+
+use crate::{
+    metrics::{ActionKind, DatabaseMetrics, LoadKind},
+    monitoring_server::MonitoringServer,
+    reporter::Reporter,
+    Diagnostics,
+};
+
+macro_rules! diagnostics_method {
+    ($(
+        pub fn $fn_name:ident(&self, $($arg_name:ident: $arg_type:ty),* $(,)?) ;
+    )*) => {
+        $(
+            pub fn $fn_name(&self, $($arg_name: $arg_type),*) {
+                self.diagnostics.$fn_name($($arg_name),*)
+            }
+        )*
+    };
+}
+
+#[derive(Debug)]
+pub struct DiagnosticsManager {
+    diagnostics: Arc<Diagnostics>,
+    reporter: Option<Reporter>,
+    monitoring_server: Option<MonitoringServer>,
+}
+
+impl DiagnosticsManager {
+    pub fn new(
+        diagnostics: Diagnostics,
+        monitoring_port: u16,
+        is_monitoring_enabled: bool,
+        is_development_mode: bool,
+    ) -> Self {
+        let deployment_id = diagnostics.server_properties.deployment_id().clone().to_owned();
+        let data_directory = diagnostics.server_metrics.data_directory().clone();
+        let is_reporting_enabled = diagnostics.server_properties.is_reporting_enabled();
+        let diagnostics = Arc::new(diagnostics);
+
+        let reporter = if is_development_mode {
+            None
+        } else {
+            Some(Reporter::new(deployment_id, diagnostics.clone(), REPORTING_URI, data_directory, is_reporting_enabled))
+        };
+
+        let monitoring_server = if is_monitoring_enabled {
+            Some(MonitoringServer::new(diagnostics.clone(), monitoring_port))
+        } else {
+            None
+        };
+
+        Self { diagnostics, reporter, monitoring_server }
+    }
+
+    diagnostics_method! {
```

Comment:
I agree that this is not ideal. However, I'd prefer to stick to having the `DiagnosticsManager` not exposing `Diagnostics` to the outside world directly. Let the `DiagnosticsManager` manage all the diagnostics-related stuff, understanding if it needs to create and manage `Reporter`s and `MonitoringServer`s, let it be ready to have additional logic for these `Diagnostics` methods in the future.

I don't want to get rid of `DiagnosticsManager`, so exposing `Diagnostics` to `Server`, `TransactionService`, and other components look verbose and incorrect. We could also potentially add other "Diagnositcs"-like things to the Manager.